src/sentry/tasks/on_demand_metrics.py

```diff
@@ -0,0 +1,370 @@
+from __future__ import annotations
+
+import logging
+from typing import Any, Optional, Sequence, Set
+
+import sentry_sdk
+from celery.exceptions import SoftTimeLimitExceeded
+
+from sentry import options
+from sentry.api.utils import get_date_range_from_params
+from sentry.models.dashboard_widget import (
+    DashboardWidgetQuery,
+    DashboardWidgetQueryOnDemand,
+    DashboardWidgetTypes,
+)
+from sentry.models.organization import Organization
+from sentry.models.project import Project
+from sentry.relay.config.metric_extraction import (
+    HashedMetricSpec,
+    convert_widget_query_to_metric,
+    on_demand_metrics_feature_flags,
+)
+from sentry.search.events import fields
+from sentry.search.events.builder import QueryBuilder
+from sentry.search.events.types import EventsResponse, QueryBuilderConfig
+from sentry.snuba.dataset import Dataset
+from sentry.snuba.referrer import Referrer
+from sentry.tasks.base import instrumented_task
+from sentry.utils import metrics
+from sentry.utils.cache import cache
+from sentry.utils.query import RangeQuerySetWrapper
+
+logger = logging.getLogger("sentry.tasks.on_demand_metrics")
+
+OnDemandExtractionState = DashboardWidgetQueryOnDemand.OnDemandExtractionState
+
+# TTL for cardinality check
+_WIDGET_QUERY_CARDINALITY_TTL = 3600 * 24 * 7  # Cardinality outcome is valid for 7 days.
+
+
+def _get_widget_processing_batch_key() -> str:
+    return "on-demand-metrics:widgets:currently-processing-batch"
+
+
+def _get_widget_query_cardinality_cache_key(widget_query: DashboardWidgetQuery) -> str:
+    return f"check-widget-query-cardinality:{widget_query.id}"
+
+
+def _set_currently_processing_batch(current_batch: int) -> None:
+    cache.set(_get_widget_processing_batch_key(), current_batch, timeout=3600)
+
+
+def _set_cardinality_cache(cache_key: str, is_low_cardinality: bool) -> None:
+    cache.set(cache_key, is_low_cardinality, timeout=_WIDGET_QUERY_CARDINALITY_TTL)
+
+
+def _get_previous_processing_batch() -> int:
+    return cache.get(_get_widget_processing_batch_key(), 0)
+
+
+def _get_current_processing_batch(total_batches: int) -> int:
+    previous_batch = _get_previous_processing_batch()
+    current_batch = (previous_batch + 1) % total_batches
+    return current_batch
+
+
+def _get_batch_for_widget_query(widget_query_id: int, total_batches: int) -> int:
+    return widget_query_id % total_batches
+
+
+class HighCardinalityWidgetException(Exception):
+    pass
+
+
+@instrumented_task(
+    name="sentry.tasks.on_demand_metrics.schedule_on_demand_check",
+    queue="on_demand_metrics",
+    max_retries=0,
+    soft_time_limit=60,
+    time_limit=120,
+    expires=60,
```

Comment:
I would increase this to 3 minutes (5 minutes - time_limit) in order to prevent running two tasks at the same time.